identity-service/src/typed-routes/solana/solanaRelayChecks.ts

```diff
@@ -0,0 +1,354 @@
+ */
+const assertAllowedRewardsManagerProgramInstruction = (
+  instructionIndex: number,
+  instruction: TransactionInstruction
+) => {
+  const decodedInstruction = decodeRewardManagerInstruction(instruction)
+  const rewardManager = decodedInstruction.keys.rewardManager.pubkey.toBase58()
+  if (rewardManager !== REWARD_MANAGER) {
+    throw new InvalidRelayInstructionError(
+      instructionIndex,
+      `Invalid Reward Manager Account: ${rewardManager}`
+    )
+  }
+}
+
+/**
+ * Checks that the claimable token authority matches the one in use on our
+ * deployed program. Optionally also check the user's social proof for transfers.
+ */
+const assertAllowedClaimableTokenProgramInstruction = async (
+  instructionIndex: number,
+  instruction: TransactionInstruction,
+  user?: { blockchainUserId?: number; handle?: string },
+  socialProofEnabled = false
+) => {
+  const decodedInstruction = decodeClaimableTokenInstruction(instruction)
+  const authority = decodedInstruction.keys.authority.pubkey
+  if (
+    !authority.equals(claimableTokenAuthorities[usdcMintAddress]) &&
+    !authority.equals(claimableTokenAuthorities[audioMintAddress])
+  ) {
+    throw new InvalidRelayInstructionError(
+      instructionIndex,
+      `Invalid Claimable Token Authority: ${authority}`
+    )
+  }
+
+  // Ensure user has social proof before withdrawing
+  // Currently disabled and removed from the relay flow
+  // TODO: Fix this so that tipping is unaffected
+  // https://linear.app/audius/issue/PAY-1941/clean-up-or-re-add-social-proof
+  if (
+    socialProofEnabled &&
+    isTransferClaimableTokenInstruction(decodedInstruction)
+  ) {
+    if (!user) {
+      throw new InvalidRelayInstructionError(
+        instructionIndex,
+        'Claimable Token Transfer requires authentication'
+      )
+    }
+    const { blockchainUserId } = user
+    const twitterUser = await (models as unknown as any).TwitterUser.findOne({
+      where: {
+        blockchainUserId
+      }
+    })
+
+    const instagramUser = await (
+      models as unknown as any
+    ).InstagramUser.findOne({
+      where: {
+        blockchainUserId
+      }
+    })
+    const hasSocialProof = !!twitterUser || !!instagramUser
+    if (!hasSocialProof) {
+      throw new InvalidRelayInstructionError(
+        instructionIndex,
+        `User is missing social proof: ${user.handle}`
+      )
+    }
+  }
+}
+
+/**
+ * Indexes of various accounts in a Jupiter V6 sharedAccountsRoute instruction.
+ * https://github.com/jup-ag/jupiter-cpi/blob/main/idl.json
+ */
+const JupiterSharedSwapAccountIndex = {
+  TOKEN_PROGRAM: 0,
+  PROGRAM_AUTHORITY: 1,
+  USER_TRANSFER_AUTHORITY: 2,
+  SOURCE_TOKEN_ACCOUNT: 3,
+  PROGRAM_SOURCE_TOKEN_ACCOUNT: 4,
+  PROGRAM_DESTINATION_TOKEN_ACCOUNT: 5,
+  DESTINATION_TOKEN_ACCOUNT: 6,
+  SOURCE_MINT: 7,
+  DESTINATION_MINT: 8,
+  PLATFORM_FEE_ACCOUNT: 9,
+  TOKEN_2022_PROGRAM: 10
+}
+const assertAllowedJupiterProgramInstruction = (
```

Comment:
trying to think thru this, noticed we are now no longer explicitly checking:
- feepayer is in set of known keys
- "destination" for rent in close instruction is our feepayer
- some stricter checks on jupiter instructions eg. `isCorrectTokenLedgerTokenAccount` in saliou's PR.
is this happening elsewhere or is there some reason we no longer need to check this?
relevant PRs:
https://github.com/AudiusProject/audius-protocol/pull/5844/files
https://github.com/AudiusProject/audius-protocol/pull/5955/files

Reply:
i see we are ignoring the token ledger instruction

Reply:
Let's chat through this live when you have a sec!

- No we're not checking feepayer, that's a good call out. Should probably check that we're not allowing swaps on the feepayer wallets themselves
- AFAIK the separate close account instruction checks should be sufficient, I don't that's relevant here (this just checks Jupiter swap instruction)
- This doesn't allow the token ledger instruction. Should I look into supporting that? From my testing it doesn't appear to be necessary for our use case.

FYI: this is a per-instruction check, not a full transaction check